src/python/pants/core/goals/deploy.py

```diff
@@ -0,0 +1,237 @@
+) -> tuple[int, tuple[str, ...]]:
+    results = []
+
+    if not process:
+        sigil = console.sigil_skipped()
+        status = "skipped"
+        if description:
+            status += f" {description}"
+        for name in names:
+            results.append(f"{sigil} {name} {status}.")
+        return 0, tuple(results)
+
+    logger.debug(f"Execute {process}")
+    res = await Effect(InteractiveProcessResult, InteractiveProcess, process)
+    if res.exit_code == 0:
+        sigil = console.sigil_succeeded()
+        status = success_status
+        prep = "to"
+    else:
+        sigil = console.sigil_failed()
+        status = "failed"
+        prep = "for"
+
+    if description:
+        status += f" {prep} {description}"
+
+    for name in names:
+        results.append(f"{sigil} {name} {status}")
+
+    return res.exit_code, tuple(results)
+
+
+@goal_rule
+async def run_deploy(console: Console, deploy_subsystem: DeploySubsystem) -> Deploy:
+    target_roots_to_deploy_field_sets = await Get(
+        TargetRootsToFieldSets,
+        TargetRootsToFieldSetsRequest(
+            DeployFieldSet,
+            goal_description=f"the `{deploy_subsystem.name}` goal",
+            no_applicable_targets_behavior=NoApplicableTargetsBehavior.error,
+        ),
+    )
+
+    deploy_processes = await MultiGet(
+        Get(DeployProcess, DeployFieldSet, field_set)
+        for field_set in target_roots_to_deploy_field_sets.field_sets
+    )
+
+    exit_code: int = 0
+    results: list[str] = []
+    published_targets: set[Target] = set()
+    for deploy in deploy_processes:
+        # If any of the deployment processes has common publish dependencies, we run the publish step only
+        # once for those common dependencies
+        publish_dependency_targets = [
+            tgt for tgt in deploy.publish_dependencies if tgt not in published_targets
+        ]
+        publish_processes = await _find_publish_processes(publish_dependency_targets)
+
+        if publish_processes:
+            logger.info(f"Publishing dependencies for {deploy.name}.")
+
+        # Publish all deployment dependencies first.
+        publish_exit_code = 0
+        for publish in publish_processes:
+            ec, statuses = await _invoke_process(
+                console,
+                publish.process,
+                names=publish.names,
+                description=publish.description,
+                success_status="published",
+            )
+            publish_exit_code = ec if ec != 0 else publish_exit_code
+            results.extend(statuses)
+
+        if publish_exit_code == 0:
+            published_targets.update(publish_dependency_targets)
+        else:
+            # If we could not publish all dependencies of a deployment process we skip the deployment step and continue.
+            logger.warning(
+                f"Failed publishing deployment dependencies for {deploy.name}. Skipping deployment."
+            )
```

Comment:
So, rather than doing this as a single loop, it would probably be cleaner to do all of the publishing ahead of time in a separate loop, and fail the run if any of the publishing failed. While we can't guarantee atomicity, avoiding partial runs of the deploy step seems like a good goal.

Reply:
When publishing everything ahead of time, if any of those publish processes fails then nothing will be deployed (as we won't move on to invoke the `DeployProcess.process`.

Happy to change it to be like that (it's a simpler implementation), I had initially decided on doing it like it is as if Pants were making a best effort to deploy something...

Reply:
My guess is that people would prefer atomicity where possible ("all or nothing")...? It might depend on whether you're in a monorepo deploying "all changed things" vs a user attempting to deploy "their two services" (for example). It's not possible to guarantee atomicity, but we can bias toward it.

Reply:
fair enough, it is also a simpler implementation so I'll amend it asap